quantum/quantum_task_state_handler.h

```diff
@@ -0,0 +1,104 @@
+};
+
+//==============================================================================================
+//                                       enum TaskState
+//==============================================================================================
+/// @enum TaskState
+/// @brief Defines the states of the task.
+/// @note Schema of the state machine: [Initialized -> Started -> *[Suspended -> Resumed] -> Stopped]
+///       Firsly, the task is created and before it's scheduled it dwells in 'Initialized' state.
+///       When the task is scheduled to be executed for the first time it switches to 'Started' state.
+///       If the task isn't completed (the execution control is passed to another coroutine) it goes into 'Suspended' state,
+///       otherwise upon the completion or termination due to an exception it switches to 'Stopped' state.
+///       If the task dwells in 'Suspended' state and scheduled for execution it goes into 'Resumed' state.
+///       The switch between 'Suspended' and 'Resumed' states may take several times until the task eventually switches over to 'Stopped' state.
+/// @note 'Suspend' and 'Resumed' states don't exist for the blocking IO task and therefore won't be handled
+/// @warning No mutexes or shared structued should ever be used as it will slow the performance significantly or
+///          grind the whole execution to a halt. Only thread local variables or pre-allocated arrays that could use
+///          the queue id as index.
+enum class TaskState : int {
+    None        = 0,      ///< No task state
+    Initialized = 1 << 0, ///< Task is initialized and isn't started yet
+    Started     = 1 << 1, ///< Task is started
+    Suspended   = 1 << 2, ///< Task isn't completed and the control of execution is passed to another coroutine
+    Resumed     = 1 << 3, ///< Task is scheduled to be executed after suspension
+    Stopped     = 1 << 4, ///< Task is completed or terminated
+    All         = Initialized |
+                  Started |
+                  Suspended |
+                  Resumed |
+                  Stopped ///< All task states
+};
+
+
+/// @brief Callback for handling task states
+/// @param[in] taskId id of the currently executing coroutine or IO task
+/// @param[in] queueId id of the queue where the task should run
+/// @param[in] state task state
+using TaskStateHandler = std::function<void(size_t taskId, int queueId, TaskState state)>;
+
+
+//==============================================================================================
+//                                    struct TaskStateConfig
+//==============================================================================================
+/// @struct TaskStateConfig
+/// @brief Contains parameters for task states handling.
+struct TaskStateConfig {
+    TaskStateHandler handler; // task state handler
+    TaskState handledStates{TaskState::None}; // task states that will be handled
```

Comment:
If a user sets handledStates to Initialized, will the callback be called? As I understand, the very first state it can ever be called for is Started?

Reply:
You are right. This `Initialized` state is required as an initial value. Otherwise, it's not correct logically  to initialize the state as `Started ` in the `Task` constructor even if it wasn't started yet